File: Neetcode/CombinationSum2.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        op = []
        def bt(i, res, cs):
            if cs == target:
                op.append(res[::])
                return
            if i >= len(candidates) or cs > target:
                return
            res.append(candidates[i])
            bt(i+1, res, cs+candidates[i])
            res.pop()
            while i+1 < len(candidates) and candidates[i+1] == candidates[i]:
                i += 1
            bt(i+1, res, cs)
        bt(0, [], 0)
        return op

File: Neetcode/CombinationSum2.cpp
class Solution {
public:
    vector<vector<int>> out;
    vector<int> res;
    void bt(vector<int> nums, int t, int i, int s){
        if (s > t || i > nums.size()){
            return;
        }
        if (s == t){
            out.push_back(res);
            return;
        }
        if (i == nums.size()){
            return;
        }
        res.push_back(nums[i]);
        bt(nums, t, i+1, s+nums[i]);
        while (i < nums.size() && nums[i] == res.back()){
            i += 1;
        }
        res.pop_back();
        bt(nums, t, i, s);
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        bt(candidates, target, 0, 0);
        return out;
    }
};
